nomad/host_volume_endpoint_test.go

```diff
@@ -0,0 +1,571 @@
+	must.NoError(t, err)
+
+	vol3, vol4 := mock.HostVolume(), mock.HostVolume()
+	vol3.NodeID = nodes[2].ID
+	vol3.NodePool = "prod"
+	vol3.Namespace = ns2
+	vol3.Name = "foobar-example"
+	vol4.Namespace = ns2
+	vol4.NodeID = nodes[1].ID
+	vol4.Name = "foobaz-example"
+
+	registerReq.Volumes = []*structs.HostVolume{vol3, vol4}
+	registerReq.Namespace = ns2
+	err = msgpackrpc.CallWithCodec(codec, "HostVolume.Register", registerReq, &registerResp)
+	must.NoError(t, err)
+
+	testCases := []struct {
+		name         string
+		req          *structs.HostVolumeListRequest
+		expectVolIDs []string
+	}{
+		{
+			name: "wrong namespace for token",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: ns1,
+					AuthToken: otherToken,
+				},
+			},
+			expectVolIDs: []string{},
+		},
+		{
+			name: "query by namespace",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: ns1,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol1.ID, vol2.ID},
+		},
+		{
+			name: "wildcard namespace",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol1.ID, vol2.ID, vol3.ID, vol4.ID},
+		},
+		{
+			name: "query by prefix",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: ns1,
+					AuthToken: token,
+					Prefix:    "foobar",
+				},
+			},
+			expectVolIDs: []string{vol1.ID},
+		},
+		{
+			name: "query by node",
+			req: &structs.HostVolumeListRequest{
+				NodeID: nodes[1].ID,
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol2.ID, vol4.ID},
+		},
+		{
+			name: "query by node pool",
+			req: &structs.HostVolumeListRequest{
+				NodePool: "prod",
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol3.ID},
+		},
+		{
+			name: "query by incompatible node ID and pool",
+			req: &structs.HostVolumeListRequest{
+				NodeID:   nodes[1].ID,
+				NodePool: "prod",
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var resp structs.HostVolumeListResponse
+			err := msgpackrpc.CallWithCodec(codec, "HostVolume.List", tc.req, &resp)
+			must.NoError(t, err)
+			gotIDs := []string{}
```

Comment:
my IDE makes orange squiggle

```suggestion
			var gotIDs []string
```

Reply:
What's the reasoning behind that? We don't lint for it.

The [short declaration](https://go.dev/ref/spec#Short_variable_declarations) is an alias functionally identical _for slices_ (or value types) to a [variable declaration](https://go.dev/ref/spec#Variable_declarations) because the result is `nil` either way and a nil slices doesn't blow up when you run built-in functions on it. But for maps and pointers those nils cause panics when you get it wrong, and being consistent between them avoids mishaps.

But it's also the wrong tool for the job here, because should be using [helper.ConvertSlice](https://github.com/hashicorp/nomad/blob/main/helper/funcs.go#L401) anyways.